File: runtime/python/prompty/prompty/jinja_subset/tokenizer.py
"""Prompty Jinja Subset — reference tokenizer.

Scans a raw template string into a flat token stream of literal text and tag
regions (``{{ }}`` expressions, ``{% %}`` statements, ``{# #}`` comments),
applying the ``{%- … -%}`` / ``{{- … -}}`` whitespace-trim markers.

This is the Phase 2 Python **reference** implementation for the Prompty Jinja
Subset (``spec/jinja-grammar.md`` §2). It is additive and intentionally NOT
registered as a renderer entry point; it exists as the conformance oracle.
"""

from __future__ import annotations

from dataclasses import dataclass

__all__ = ["Token", "tokenize", "TemplateSyntaxError"]


class TemplateSyntaxError(ValueError):
    """Raised when the template cannot be tokenized or parsed."""


@dataclass
class Token:
    """A single lexical token.

    ``type`` is one of ``"text"``, ``"expr"``, ``"stmt"``. Comment tokens are
    dropped during tokenization (after their trim markers are applied to
    neighbours). ``value`` is the raw text for ``"text"`` tokens or the inner
    (delimiter- and trim-marker-stripped) source for tag tokens.
    """

    type: str
    value: str
    trim_left: bool = False
    trim_right: bool = False


_OPENERS = {"{{": ("expr", "}}"), "{%": ("stmt", "%}"), "{#": ("comment", "#}")}


def tokenize(template: str) -> list[Token]:
    """Tokenize ``template`` into text/expr/stmt tokens with trims applied."""
    raw: list[Token] = []
    i = 0
    n = len(template)
    text_start = 0

    while i < n:
        two = template[i : i + 2]
        if two in _OPENERS:
            # Flush any pending literal text.
            if i > text_start:
                raw.append(Token("text", template[text_start:i]))
            kind, close = _OPENERS[two]
            close_idx = template.find(close, i + 2)
            if close_idx == -1:
                raise TemplateSyntaxError(f"Unclosed '{two}' tag at offset {i}")
            inner = template[i + 2 : close_idx]
            trim_left = inner.startswith("-")
            trim_right = inner.endswith("-")
            if trim_left:
                inner = inner[1:]
            if trim_right:
                inner = inner[:-1]
            if kind != "comment":
                raw.append(Token(kind, inner.strip(), trim_left, trim_right))
            else:
                # Comment produces no node but still carries trim semantics.
                raw.append(Token("comment", "", trim_left, trim_right))
            i = close_idx + len(close)
            text_start = i
        else:
            i += 1

    if text_start < n:
        raw.append(Token("text", template[text_start:]))

    _apply_trims(raw)
    return [t for t in raw if t.type != "comment"]


def _apply_trims(tokens: list[Token]) -> None:
    """Apply ``-`` trim markers in place, stripping neighbouring whitespace."""
    for idx, tok in enumerate(tokens):
        if tok.type == "text":
            continue
        if tok.trim_left and idx > 0 and tokens[idx - 1].type == "text":
            tokens[idx - 1].value = tokens[idx - 1].value.rstrip()
        if tok.trim_right and idx + 1 < len(tokens) and tokens[idx + 1].type == "text":
            tokens[idx + 1].value = tokens[idx + 1].value.lstrip()

File: runtime/java/prompty/src/main/java/com/microsoft/prompty/jinjasubset/Tokenizer.java
package com.microsoft.prompty.jinjasubset;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

final class Tokenizer {
  enum Type { TEXT, EXPR, STMT, COMMENT }

  static final class Token {
    final Type type;
    String value;
    final boolean trimLeft;
    final boolean trimRight;

    Token(Type type, String value) { this(type, value, false, false); }

    Token(Type type, String value, boolean trimLeft, boolean trimRight) {
      this.type = type;
      this.value = value;
      this.trimLeft = trimLeft;
      this.trimRight = trimRight;
    }
  }

  private record Opener(Type type, String close) {}

  private static final Map<String, Opener> OPENERS = new HashMap<>();

  static {
    OPENERS.put("{{", new Opener(Type.EXPR, "}}"));
    OPENERS.put("{%", new Opener(Type.STMT, "%}"));
    OPENERS.put("{#", new Opener(Type.COMMENT, "#}"));
  }

  private Tokenizer() {}

  static List<Token> tokenize(String template) {
    String src = template == null ? "" : template;
    List<Token> raw = new ArrayList<>();
    int i = 0;
    int textStart = 0;
    while (i < src.length()) {
      String two = i + 2 <= src.length() ? src.substring(i, i + 2) : "";
      Opener opener = OPENERS.get(two);
      if (opener != null) {
        if (i > textStart) {
          raw.add(new Token(Type.TEXT, src.substring(textStart, i)));
        }
        int closeIdx = src.indexOf(opener.close(), i + 2);
        if (closeIdx < 0) {
          throw new TemplateSyntaxException("Unclosed '" + two + "' tag at offset " + i);
        }
        String inner = src.substring(i + 2, closeIdx);
        boolean trimLeft = inner.startsWith("-");
        boolean trimRight = inner.endsWith("-");
        if (trimLeft) inner = inner.substring(1);
        if (trimRight) inner = inner.substring(0, inner.length() - 1);
        raw.add(new Token(opener.type(), opener.type() == Type.COMMENT ? "" : inner.trim(), trimLeft, trimRight));
        i = closeIdx + opener.close().length();
        textStart = i;
      } else {
        i++;
      }
    }
    if (textStart < src.length()) {
      raw.add(new Token(Type.TEXT, src.substring(textStart)));
    }
    applyTrims(raw);
    List<Token> result = new ArrayList<>();
    for (Token token : raw) {
      if (token.type != Type.COMMENT) result.add(token);
    }
    return result;
  }

  private static void applyTrims(List<Token> tokens) {
    for (int i = 0; i < tokens.size(); i++) {
      Token token = tokens.get(i);
      if (token.type == Type.TEXT) continue;
      if (token.trimLeft && i > 0 && tokens.get(i - 1).type == Type.TEXT) {
        tokens.get(i - 1).value = rstrip(tokens.get(i - 1).value);
      }
      if (token.trimRight && i + 1 < tokens.size() && tokens.get(i + 1).type == Type.TEXT) {
        tokens.get(i + 1).value = lstrip(tokens.get(i + 1).value);
      }
    }
  }

  private static String lstrip(String value) {
    int i = 0;
    while (i < value.length() && Character.isWhitespace(value.charAt(i))) i++;
    return value.substring(i);
  }

  private static String rstrip(String value) {
    int i = value.length();
    while (i > 0 && Character.isWhitespace(value.charAt(i - 1))) i--;
    return value.substring(0, i);
  }
}
